packages/react-sdk/src/db/instance.ts

```diff
@@ -32,3 +32,90 @@ export const Instance = z.lazy(
     })
   // @todo can't figure out how to make component to be z.enum(Object.keys(components))
 ) as z.ZodType<Instance>;
+
+export type NormalizedInstance = {
+  id: string;
+  component: ComponentName;
+  cssRules: Array<CssRule>;
+  /**
+   * number is an index in normalized tree
+   * string is text content
+   */
+  children: (number | string)[];
+};
+
+export type NormalizedTree = (undefined | NormalizedInstance)[];
+
+export const normalizeTree = (rootInstance: Instance) => {
+  const normalizedTree: NormalizedTree = [];
+  const traverse = (
+    instance: Instance,
+    normalizedInstance: NormalizedInstance
+  ) => {
+    for (const child of instance.children) {
+      if (typeof child === "string") {
+        normalizedInstance.children.push(child);
+      } else {
+        const normalizedChild = {
+          id: child.id,
+          component: child.component,
+          cssRules: child.cssRules,
+          children: [],
+        };
+        // push returns length
+        const index = normalizedTree.push(normalizedChild) - 1;
+        normalizedInstance.children.push(index);
```

Comment:
we need to deeply discuss consequences of using index instead of ids

Reply:
It was a fair point about conflicts between clients, ids will work better.